message_passing/message_handler_waiting_compensation_cleanup/workflows.py

```diff
@@ -0,0 +1,121 @@
+import asyncio
+from datetime import timedelta
+
+from temporalio import exceptions, workflow
+
+from message_passing.message_handler_waiting_compensation_cleanup import (
+    OnWorkflowExitAction,
+    UpdateInput,
+    WorkflowExitType,
+    WorkflowInput,
+)
+from message_passing.message_handler_waiting_compensation_cleanup.activities import (
+    my_activity,
+)
+
+
+@workflow.defn
+class MyWorkflow:
+    """
+    This Workflow upholds the following recommended practices:
+
+    1. The main workflow method ensures that all signal and update handlers are
+       finished before a successful return, and on failure, cancellation, and
+       continue-as-new.
+    2. The update handler performs any necessary compensation/cleanup when the
+       workflow is cancelled, fails, or continues-as-new.
+    """
+
+    def __init__(self) -> None:
+        self.workflow_exit_exception: asyncio.Future[BaseException] = asyncio.Future()
+
+    @workflow.run
+    async def run(self, input: WorkflowInput) -> str:
+        try:
+            # 👉 Use this `try...except` style, instead of waiting for message
+            # handlers to finish in a `finally` block. The reason is that other
+            # exception types will cause a Workflow Task failure, in which case
+            # we do *not* want to wait for message handlers to finish.
+            result = await self._run(input)
```

Comment:
I think splitting the run logic to half at the top of the class and have way at the bottom makes the sample hard to read, consider inlining

Reply:
Arguably, it's just a bit hard to follow to have to jump around the file for this very simple primary workflow logic. But not a big deal.